### ISSUE TITLE
Inconsistent behavior returning [self]

### ISSUE THREAD

**Issue by jtackaberry** [microsoft/pyright#5784]
Returning `[self]` yields different results depending on whether it's returned directly or via an intermediate variable.

**Describe the bug**

Jumping right into the code:

```python
class Foo:
    # This is fine
    def return_self(self) -> 'Foo':
        return self
    
    # So is this
    def return_list_of_self(self) -> list['Foo']:
        return [self]
    
    # This errors:
    # Pyright: Expression of type "list[Self@Foo]" cannot be assigned to return type "list[Foo]"
    #  "list[Self@Foo]" is incompatible with "list[Foo]"
    #    Type parameter "_T@list" is invariant, but "Self@Foo" is not the same as "Foo" [reportGeneralTypeIssues]
    def return_list_of_self_with_intermediate_variable(self) -> list['Foo']:
        r = [self]
        return r
```

I had expected the last case to work properly, but in the process of trying to better understand the error pyright was reporting, I tested with the second method and was surprised to see the behavior was different.

Even if the error is legitimate (in which case I have some work to do to understand why), shouldn't the last two methods at least produce the same results with pyright?

**VS Code extension or command-line**

Using pyright 1.1.323 with neovim 0.9.1 on Python 3.10.12.

**Comment by erictraut** [comment]
Pyright's behavior is correct here. Here's a simpler example that demonstrates the difference between the two methods.

```python
v1 = [1, 2, 3] # Inferred type of v1 is `list[int]`
x1: list[float] = v1 # Type violation

x2: list[float] = [1, 2, 3] # No error because of bidirectional type inference
```

If you use an intermediate variable `r`, the inferred type of `[self]` is `list[Self]`. The type `list[Self]` is not a subtype of `list[Foo]` because lists are invariant. 

You can fix this by adding a type declaration for `r`.

```python
r: list[Foo] = [self]
```

For more details, refer to the documentation for [type inference rules for lists](https://microsoft.github.io/pyright/#/type-inference?id=list-expressions) and [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types).

**Comment by jtackaberry** [comment]
Thanks @erictraut for the clear explanation.  Seems obvious now in hindsight. :)
